// @wordpress/components/build-module/toolbar-button/index.js
import _extends from "@babel/runtime/helpers/esm/extends";
import { createElement } from "@wordpress/element";

/**
 * External dependencies
 */
import classnames from 'classnames';
/**
 * WordPress dependencies
 */

import { useContext } from '@wordpress/element';
/**
 * Internal dependencies
 */

import IconButton from '../icon-button';
import ToolbarContext from '../toolbar-context';
import AccessibleToolbarButtonContainer from './accessible-toolbar-button-container';
import ToolbarButtonContainer from './toolbar-button-container';

function ToolbarButton(_ref) {
  var containerClassName = _ref.containerClassName,
      icon = _ref.icon,
      title = _ref.title,
      shortcut = _ref.shortcut,
      subscript = _ref.subscript,
      _onClick = _ref.onClick,
      className = _ref.className,
      isActive = _ref.isActive,
      isDisabled = _ref.isDisabled,
      extraProps = _ref.extraProps,
      children = _ref.children;
  // It'll contain state if `ToolbarButton` is being used within
  // `<Toolbar __experimentalAccessibilityLabel="label" />`
  var accessibleToolbarState = useContext(ToolbarContext);
  var button = createElement(IconButton, _extends({
    icon: icon,
    label: title,
    shortcut: shortcut,
    "data-subscript": subscript,
    onClick: function onClick(event) {
      event.stopPropagation();

      if (_onClick) {
        _onClick(event);
      }
    },
    className: classnames('components-toolbar__control', className, {
      'is-active': isActive
    }),
    "aria-pressed": isActive,
    disabled: isDisabled
  }, extraProps));

  if (accessibleToolbarState) {
    return createElement(AccessibleToolbarButtonContainer, {
      className: containerClassName
    }, button);
  } // ToolbarButton is being used outside of the accessible Toolbar


  return createElement(ToolbarButtonContainer, {
    className: containerClassName
  }, button, children);
}

export default ToolbarButton;
