// @wordpress/block-library/build/navigation/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.settings = exports.name = void 0;

var _i18n = require("@wordpress/i18n");

var _icons = require("@wordpress/icons");

var _edit = _interopRequireDefault(require("./edit"));

var _save = _interopRequireDefault(require("./save"));

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
var name = 'core/navigation';
exports.name = name;
var settings = {
  title: (0, _i18n.__)('Navigation'),
  icon: _icons.menu,
  description: (0, _i18n.__)('Add a navigation block to your site.'),
  keywords: [(0, _i18n.__)('menu'), (0, _i18n.__)('navigation'), (0, _i18n.__)('links')],
  category: 'layout',
  supports: {
    align: ['wide', 'full'],
    anchor: true,
    html: false,
    inserter: true
  },
  styles: [{
    name: 'light',
    label: (0, _i18n.__)('Light'),
    isDefault: true
  }, {
    name: 'dark',
    label: (0, _i18n.__)('Dark')
  }],
  edit: _edit.default,
  save: _save.default
};
exports.settings = settings;
