// @wordpress/block-library/build/navigation/use-block-navigator.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = useBlockNavigator;

var _element = require("@wordpress/element");

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _components = require("@wordpress/components");

var _i18n = require("@wordpress/i18n");

var _blockNavigationList = _interopRequireDefault(require("./block-navigation-list"));

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
var NavigatorIcon = (0, _element.createElement)(_components.SVG, {
  xmlns: "http://www.w3.org/2000/svg",
  viewBox: "0 0 24 24",
  width: "20",
  height: "20"
}, (0, _element.createElement)(_components.Path, {
  d: "M5 5H3v2h2V5zm3 8h11v-2H8v2zm9-8H6v2h11V5zM7 11H5v2h2v-2zm0 8h2v-2H7v2zm3-2v2h11v-2H10z"
}));

function useBlockNavigator(clientId) {
  var _useState = (0, _element.useState)(false),
      _useState2 = (0, _slicedToArray2.default)(_useState, 2),
      isNavigationListOpen = _useState2[0],
      setIsNavigationListOpen = _useState2[1];

  var navigatorToolbarButton = (0, _element.createElement)(_components.Button, {
    className: "components-toolbar__control",
    label: (0, _i18n.__)('Open block navigator'),
    onClick: function onClick() {
      return setIsNavigationListOpen(true);
    },
    icon: NavigatorIcon
  });
  var navigatorModal = isNavigationListOpen && (0, _element.createElement)(_components.Modal, {
    title: (0, _i18n.__)('Block Navigator'),
    closeLabel: (0, _i18n.__)('Close'),
    onRequestClose: function onRequestClose() {
      setIsNavigationListOpen(false);
    }
  }, (0, _element.createElement)(_blockNavigationList.default, {
    clientId: clientId
  }));
  return {
    navigatorToolbarButton: navigatorToolbarButton,
    navigatorModal: navigatorModal
  };
}
