// @wordpress/components/build-module/card/header.js
import _extends from "@babel/runtime/helpers/esm/extends";
import _objectSpread from "@babel/runtime/helpers/esm/objectSpread";
import _objectWithoutProperties from "@babel/runtime/helpers/esm/objectWithoutProperties";
import { createElement } from "@wordpress/element";

/**
 * External dependencies
 */
import classnames from 'classnames';
/**
 * Internal dependencies
 */

import { HeaderUI } from './styles/card-styles';
import { useCardContext } from './context';
export var defaultProps = {
  isBorderless: false,
  isShady: false,
  size: 'medium'
};
export function CardHeader(props) {
  var className = props.className,
      isShady = props.isShady,
      additionalProps = _objectWithoutProperties(props, ["className", "isShady"]);

  var mergedProps = _objectSpread({}, defaultProps, useCardContext(), props);

  var isBorderless = mergedProps.isBorderless,
      size = mergedProps.size;
  var classes = classnames('components-card__header', isBorderless && 'is-borderless', isShady && 'is-shady', size && "is-size-".concat(size), className);
  return createElement(HeaderUI, _extends({}, additionalProps, {
    className: classes
  }));
}
export default CardHeader;
