// @wordpress/block-library/build/social-link/social-list.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getNameBySite = exports.getIconBySite = exports.default = void 0;

var _icons = require("./icons");

/**
 * Internal dependencies
 */
var socialList = {
  fivehundredpx: {
    name: '500px',
    icon: _icons.FivehundredpxIcon
  },
  amazon: {
    name: 'Amazon',
    icon: _icons.AmazonIcon
  },
  bandcamp: {
    name: 'Bandcamp',
    icon: _icons.BandcampIcon
  },
  behance: {
    name: 'Behance',
    icon: _icons.BehanceIcon
  },
  chain: {
    name: 'Link',
    icon: _icons.ChainIcon
  },
  codepen: {
    name: 'CodePen',
    icon: _icons.CodepenIcon
  },
  deviantart: {
    name: 'DeviantArt',
    icon: _icons.DeviantArtIcon
  },
  dribbble: {
    name: 'Dribbble',
    icon: _icons.DribbbleIcon
  },
  dropbox: {
    name: 'Dropbox',
    icon: _icons.DropboxIcon
  },
  etsy: {
    name: 'Etsy',
    icon: _icons.EtsyIcon
  },
  facebook: {
    name: 'Facebook',
    icon: _icons.FacebookIcon
  },
  feed: {
    name: 'RSS Feed',
    icon: _icons.FeedIcon
  },
  flickr: {
    name: 'Flickr',
    icon: _icons.FlickrIcon
  },
  foursquare: {
    name: 'Foursquare',
    icon: _icons.FoursquareIcon
  },
  goodreads: {
    name: 'Goodreads',
    icon: _icons.GoodreadsIcon
  },
  google: {
    name: 'Google',
    icon: _icons.GoogleIcon
  },
  github: {
    name: 'GitHub',
    icon: _icons.GitHubIcon
  },
  instagram: {
    name: 'Instagram',
    icon: _icons.InstagramIcon
  },
  lastfm: {
    name: 'Last.fm',
    icon: _icons.LastfmIcon
  },
  linkedin: {
    name: 'LinkedIn',
    icon: _icons.LinkedInIcon
  },
  mail: {
    name: 'Mail',
    icon: _icons.MailIcon
  },
  mastodon: {
    name: 'Mastodon',
    icon: _icons.MastodonIcon
  },
  meetup: {
    name: 'Meetup',
    icon: _icons.MeetupIcon
  },
  medium: {
    name: 'Medium',
    icon: _icons.MediumIcon
  },
  pinterest: {
    name: 'Pinterest',
    icon: _icons.PinterestIcon
  },
  pocket: {
    name: 'Pocket',
    icon: _icons.PocketIcon
  },
  reddit: {
    name: 'Reddit',
    icon: _icons.RedditIcon
  },
  skype: {
    name: 'Skype',
    icon: _icons.SkypeIcon
  },
  snapchat: {
    name: 'Snapchat',
    icon: _icons.SnapchatIcon
  },
  soundcloud: {
    name: 'SoundCloud',
    icon: _icons.SoundCloudIcon
  },
  spotify: {
    name: 'Spotify',
    icon: _icons.SpotifyIcon
  },
  tumblr: {
    name: 'Tumblr',
    icon: _icons.TumblrIcon
  },
  twitch: {
    name: 'Twitch',
    icon: _icons.TwitchIcon
  },
  twitter: {
    name: 'Twitter',
    icon: _icons.TwitterIcon
  },
  vimeo: {
    name: 'Vimeo',
    icon: _icons.VimeoIcon
  },
  vk: {
    name: 'VK',
    icon: _icons.VkIcon
  },
  wordpress: {
    name: 'WordPress',
    icon: _icons.WordPressIcon
  },
  yelp: {
    name: 'Yelp',
    icon: _icons.YelpIcon
  },
  youtube: {
    name: 'YouTube',
    icon: _icons.YouTubeIcon
  }
};
var _default = socialList;
/**
 * Retrieves the social service's icon component.
 *
 * @param {string} site key for a social service (lowercase slug)
 *
 * @return {WPComponent} Icon component for social service.
 */

exports.default = _default;

var getIconBySite = function getIconBySite(site) {
  return socialList[site].icon;
};
/**
 * Retrieves the display name for the social service.
 *
 * @param {string} site key for a social service (lowercase slug)
 *
 * @return {string} Display name for social service
 */


exports.getIconBySite = getIconBySite;

var getNameBySite = function getNameBySite(site) {
  return socialList[site].name;
};

exports.getNameBySite = getNameBySite;
