// @wordpress/components/build/toolbar-button/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _classnames = _interopRequireDefault(require("classnames"));

var _iconButton = _interopRequireDefault(require("../icon-button"));

var _toolbarContext = _interopRequireDefault(require("../toolbar-context"));

var _accessibleToolbarButtonContainer = _interopRequireDefault(require("./accessible-toolbar-button-container"));

var _toolbarButtonContainer = _interopRequireDefault(require("./toolbar-button-container"));

/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
function ToolbarButton(_ref) {
  var containerClassName = _ref.containerClassName,
      icon = _ref.icon,
      title = _ref.title,
      shortcut = _ref.shortcut,
      subscript = _ref.subscript,
      _onClick = _ref.onClick,
      className = _ref.className,
      isActive = _ref.isActive,
      isDisabled = _ref.isDisabled,
      extraProps = _ref.extraProps,
      children = _ref.children;
  // It'll contain state if `ToolbarButton` is being used within
  // `<Toolbar __experimentalAccessibilityLabel="label" />`
  var accessibleToolbarState = (0, _element.useContext)(_toolbarContext.default);
  var button = (0, _element.createElement)(_iconButton.default, (0, _extends2.default)({
    icon: icon,
    label: title,
    shortcut: shortcut,
    "data-subscript": subscript,
    onClick: function onClick(event) {
      event.stopPropagation();

      if (_onClick) {
        _onClick(event);
      }
    },
    className: (0, _classnames.default)('components-toolbar__control', className, {
      'is-active': isActive
    }),
    "aria-pressed": isActive,
    disabled: isDisabled
  }, extraProps));

  if (accessibleToolbarState) {
    return (0, _element.createElement)(_accessibleToolbarButtonContainer.default, {
      className: containerClassName
    }, button);
  } // ToolbarButton is being used outside of the accessible Toolbar


  return (0, _element.createElement)(_toolbarButtonContainer.default, {
    className: containerClassName
  }, button, children);
}

var _default = ToolbarButton;
exports.default = _default;
