// @wordpress/block-library/build-module/navigation-link/edit.js
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";
import { createElement } from "@wordpress/element";

/**
 * External dependencies
 */
import classnames from 'classnames';
import { escape, unescape } from 'lodash';
/**
 * WordPress dependencies
 */

import { compose } from '@wordpress/compose';
import { createBlock } from '@wordpress/blocks';
import { withDispatch, withSelect } from '@wordpress/data';
import { ExternalLink, KeyboardShortcuts, PanelBody, Popover, TextareaControl, TextControl, ToggleControl, ToolbarButton, ToolbarGroup } from '@wordpress/components';
import { rawShortcut, displayShortcut } from '@wordpress/keycodes';
import { __ } from '@wordpress/i18n';
import { BlockControls, InnerBlocks, InspectorControls, RichText, __experimentalLinkControl as LinkControl } from '@wordpress/block-editor';
import { isURL, prependHTTP } from '@wordpress/url';
import { Fragment, useState, useEffect, useRef } from '@wordpress/element';
import { placeCaretAtHorizontalEdge } from '@wordpress/dom';
/**
 * Internal dependencies
 */

import { toolbarSubmenuIcon, itemSubmenuIcon } from './icons';

function NavigationLinkEdit(_ref) {
  var attributes = _ref.attributes,
      hasDescendants = _ref.hasDescendants,
      isSelected = _ref.isSelected,
      isParentOfSelectedBlock = _ref.isParentOfSelectedBlock,
      setAttributes = _ref.setAttributes,
      showSubmenuIcon = _ref.showSubmenuIcon,
      insertLinkBlock = _ref.insertLinkBlock;
  var label = attributes.label,
      opensInNewTab = attributes.opensInNewTab,
      title = attributes.title,
      url = attributes.url,
      nofollow = attributes.nofollow,
      description = attributes.description;
  var link = {
    title: title ? unescape(title) : '',
    url: url,
    opensInNewTab: opensInNewTab
  };

  var _useState = useState(false),
      _useState2 = _slicedToArray(_useState, 2),
      isLinkOpen = _useState2[0],
      setIsLinkOpen = _useState2[1];

  var itemLabelPlaceholder = __('Add link…');

  var ref = useRef(); // Show the LinkControl on mount if the URL is empty
  // ( When adding a new menu item)
  // This can't be done in the useState call because it cconflicts
  // with the autofocus behavior of the BlockListBlock component.

  useEffect(function () {
    if (!url) {
      setIsLinkOpen(true);
    }
  }, []);
  /**
   * The hook shouldn't be necessary but due to a focus loss happening
   * when selecting a suggestion in the link popover, we force close on block unselection.
   */

  useEffect(function () {
    if (!isSelected) {
      setIsLinkOpen(false);
    }
  }, [isSelected]); // If the LinkControl popover is open and the URL has changed, close the LinkControl and focus the label text.

  useEffect(function () {
    if (isLinkOpen && url) {
      // Close the link.
      setIsLinkOpen(false); // Does this look like a URL and have something TLD-ish?

      if (isURL(prependHTTP(label)) && /^.+\.[a-z]+/.test(label)) {
        // Focus and select the label text.
        selectLabelText();
      } else {
        // Focus it (but do not select).
        placeCaretAtHorizontalEdge(ref.current, true);
      }
    }
  }, [url]);
  /**
   * Focus the navigation link label text and select it.
   */

  function selectLabelText() {
    ref.current.focus();
    var selection = window.getSelection();
    var range = document.createRange(); // Get the range of the current ref contents so we can add this range to the selection.

    range.selectNodeContents(ref.current);
    selection.removeAllRanges();
    selection.addRange(range);
  }

  return createElement(Fragment, null, createElement(BlockControls, null, createElement(ToolbarGroup, null, createElement(KeyboardShortcuts, {
    bindGlobal: true,
    shortcuts: _defineProperty({}, rawShortcut.primary('k'), function () {
      return setIsLinkOpen(true);
    })
  }), createElement(ToolbarButton, {
    name: "link",
    icon: "admin-links",
    title: __('Link'),
    shortcut: displayShortcut.primary('k'),
    onClick: function onClick() {
      return setIsLinkOpen(true);
    }
  }), createElement(ToolbarButton, {
    name: "submenu",
    icon: toolbarSubmenuIcon,
    title: __('Add submenu'),
    onClick: insertLinkBlock
  }))), createElement(InspectorControls, null, createElement(PanelBody, {
    title: __('SEO settings')
  }, createElement(TextControl, {
    value: title || '',
    onChange: function onChange(titleValue) {
      setAttributes({
        title: titleValue
      });
    },
    label: __('Title Attribute'),
    help: __('Provide more context about where the link goes.')
  }), createElement(ToggleControl, {
    checked: nofollow,
    onChange: function onChange(nofollowValue) {
      setAttributes({
        nofollow: nofollowValue
      });
    },
    label: __('Add nofollow to link'),
    help: createElement(Fragment, null, __("Don't let search engines follow this link."), createElement(ExternalLink, {
      className: "wp-block-navigation-link__nofollow-external-link",
      href: __('https://codex.wordpress.org/Nofollow')
    }, __("What's this?")))
  })), createElement(PanelBody, {
    title: __('Link settings')
  }, createElement(TextareaControl, {
    value: description || '',
    onChange: function onChange(descriptionValue) {
      setAttributes({
        description: descriptionValue
      });
    },
    label: __('Description'),
    help: __('The description will be displayed in the menu if the current theme supports it.')
  }))), createElement("div", {
    className: classnames('wp-block-navigation-link', {
      'is-editing': isSelected || isParentOfSelectedBlock,
      'is-selected': isSelected,
      'has-link': !!url
    })
  }, createElement("div", {
    className: "wp-block-navigation-link__content"
  }, createElement(RichText, {
    ref: ref,
    tagName: "span",
    className: "wp-block-navigation-link__label",
    value: label,
    onChange: function onChange(labelValue) {
      return setAttributes({
        label: labelValue
      });
    },
    placeholder: itemLabelPlaceholder,
    withoutInteractiveFormatting: true,
    allowedFormats: ['core/bold', 'core/italic', 'core/image', 'core/strikethrough']
  }), showSubmenuIcon && createElement("span", {
    className: "wp-block-navigation-link__submenu-icon"
  }, itemSubmenuIcon), isLinkOpen && createElement(Popover, {
    position: "bottom center",
    onClose: function onClose() {
      return setIsLinkOpen(false);
    }
  }, createElement(LinkControl, {
    className: "wp-block-navigation-link__inline-link-input",
    value: link,
    showInitialSuggestions: true,
    onChange: function onChange() {
      var _ref3 = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {},
          _ref3$title = _ref3.title,
          newTitle = _ref3$title === void 0 ? '' : _ref3$title,
          _ref3$url = _ref3.url,
          newURL = _ref3$url === void 0 ? '' : _ref3$url,
          newOpensInNewTab = _ref3.opensInNewTab,
          id = _ref3.id;

      return setAttributes({
        title: escape(newTitle),
        url: encodeURI(newURL),
        label: function () {
          var normalizedTitle = newTitle.replace(/http(s?):\/\//gi, '');
          var normalizedURL = newURL.replace(/http(s?):\/\//gi, '');

          if (newTitle !== '' && normalizedTitle !== normalizedURL && label !== newTitle) {
            return escape(newTitle);
          } else if (label) {
            return label;
          } // If there's no label, add the URL.


          return escape(normalizedURL);
        }(),
        opensInNewTab: newOpensInNewTab,
        id: id
      });
    }
  }))), createElement(InnerBlocks, {
    allowedBlocks: ['core/navigation-link'],
    renderAppender: hasDescendants && isSelected || isParentOfSelectedBlock ? InnerBlocks.DefaultAppender : false
  })));
}

export default compose([withSelect(function (select, ownProps) {
  var _select = select('core/block-editor'),
      getBlockName = _select.getBlockName,
      getBlockAttributes = _select.getBlockAttributes,
      getBlockParents = _select.getBlockParents,
      getClientIdsOfDescendants = _select.getClientIdsOfDescendants,
      hasSelectedInnerBlock = _select.hasSelectedInnerBlock;

  var clientId = ownProps.clientId;
  var rootBlock = getBlockParents(clientId)[0];
  var parentBlock = getBlockParents(clientId, true)[0];
  var rootBlockAttributes = getBlockAttributes(rootBlock);
  var hasDescendants = !!getClientIdsOfDescendants([clientId]).length;
  var isLevelZero = getBlockName(parentBlock) === 'core/navigation';
  var showSubmenuIcon = rootBlockAttributes.showSubmenuIcon && isLevelZero && hasDescendants;
  var isParentOfSelectedBlock = hasSelectedInnerBlock(clientId, true);
  return {
    isParentOfSelectedBlock: isParentOfSelectedBlock,
    hasDescendants: hasDescendants,
    showSubmenuIcon: showSubmenuIcon
  };
}), withDispatch(function (dispatch, ownProps, registry) {
  return {
    insertLinkBlock: function insertLinkBlock() {
      var clientId = ownProps.clientId;

      var _dispatch = dispatch('core/block-editor'),
          insertBlock = _dispatch.insertBlock;

      var _registry$select = registry.select('core/block-editor'),
          getClientIdsOfDescendants = _registry$select.getClientIdsOfDescendants;

      var navItems = getClientIdsOfDescendants([clientId]);
      var insertionPoint = navItems.length ? navItems.length : 0;
      var blockToInsert = createBlock('core/navigation-link');
      insertBlock(blockToInsert, insertionPoint, clientId);
    }
  };
})])(NavigationLinkEdit);
