// @wordpress/block-library/build-module/navigation-link/index.js
/**
 * WordPress dependencies
 */
import { __ } from '@wordpress/i18n';
import { navigation as icon } from '@wordpress/icons';
/**
 * Internal dependencies
 */

var metadata = {
  name: "core/navigation-link",
  category: "layout",
  attributes: {
    label: {
      type: "string"
    },
    nofollow: {
      type: "boolean",
      "default": false
    },
    title: {
      type: "string"
    },
    type: {
      type: "string"
    },
    description: {
      type: "string"
    },
    id: {
      type: "number"
    },
    opensInNewTab: {
      type: "boolean",
      "default": false
    },
    url: {
      type: "string"
    }
  }
};
import edit from './edit';
import save from './save';
var name = metadata.name;
export { metadata, name };
export var settings = {
  title: __('Navigation Link'),
  parent: ['core/navigation'],
  icon: icon,
  description: __('Add a page, link, or another item to your navigation.'),
  supports: {
    reusable: false,
    html: false
  },
  __experimentalLabel: function __experimentalLabel(_ref) {
    var label = _ref.label;
    return label;
  },
  edit: edit,
  save: save
};
