// @wordpress/components/build/card/footer.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CardFooter = CardFooter;
exports.default = exports.defaultProps = void 0;

var _element = require("@wordpress/element");

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _classnames = _interopRequireDefault(require("classnames"));

var _cardStyles = require("./styles/card-styles");

var _context = require("./context");

/**
 * External dependencies
 */

/**
 * Internal dependencies
 */
var defaultProps = {
  isBorderless: false,
  isShady: false,
  size: 'medium'
};
exports.defaultProps = defaultProps;

function CardFooter(props) {
  var className = props.className,
      isShady = props.isShady,
      additionalProps = (0, _objectWithoutProperties2.default)(props, ["className", "isShady"]);
  var mergedProps = (0, _objectSpread2.default)({}, defaultProps, (0, _context.useCardContext)(), props);
  var isBorderless = mergedProps.isBorderless,
      size = mergedProps.size;
  var classes = (0, _classnames.default)('components-card__footer', isBorderless && 'is-borderless', isShady && 'is-shady', size && "is-size-".concat(size), className);
  return (0, _element.createElement)(_cardStyles.FooterUI, (0, _extends2.default)({}, additionalProps, {
    className: classes
  }));
}

var _default = CardFooter;
exports.default = _default;
