// @wordpress/block-library/build/navigation-link/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.settings = exports.name = exports.metadata = void 0;

var _i18n = require("@wordpress/i18n");

var _icons = require("@wordpress/icons");

var _edit = _interopRequireDefault(require("./edit"));

var _save = _interopRequireDefault(require("./save"));

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
var metadata = {
  name: "core/navigation-link",
  category: "layout",
  attributes: {
    label: {
      type: "string"
    },
    nofollow: {
      type: "boolean",
      "default": false
    },
    title: {
      type: "string"
    },
    type: {
      type: "string"
    },
    description: {
      type: "string"
    },
    id: {
      type: "number"
    },
    opensInNewTab: {
      type: "boolean",
      "default": false
    },
    url: {
      type: "string"
    }
  }
};
exports.metadata = metadata;
var name = metadata.name;
exports.name = name;
var settings = {
  title: (0, _i18n.__)('Navigation Link'),
  parent: ['core/navigation'],
  icon: _icons.navigation,
  description: (0, _i18n.__)('Add a page, link, or another item to your navigation.'),
  supports: {
    reusable: false,
    html: false
  },
  __experimentalLabel: function __experimentalLabel(_ref) {
    var label = _ref.label;
    return label;
  },
  edit: _edit.default,
  save: _save.default
};
exports.settings = settings;
