// @wordpress/components/build-module/button/index.native.js
import _objectSpread from "@babel/runtime/helpers/esm/objectSpread";
import { createElement } from "@wordpress/element";

/**
 * External dependencies
 */
import { StyleSheet, TouchableOpacity, Text, View, Platform } from 'react-native';
/**
 * WordPress dependencies
 */

import { Children, cloneElement } from '@wordpress/element';
import { withPreferredColorScheme } from '@wordpress/compose';
var isAndroid = Platform.OS === 'android';
var marginBottom = isAndroid ? -0.5 : 0;
var marginLeft = -3;
var styles = StyleSheet.create({
  container: {
    flex: 1,
    padding: 3,
    justifyContent: 'center',
    alignItems: 'center'
  },
  buttonInactive: {
    flex: 1,
    flexDirection: 'row',
    justifyContent: 'center',
    alignItems: 'center'
  },
  fixedRatio: {
    aspectRatio: 1
  },
  buttonActive: {
    flex: 1,
    flexDirection: 'row',
    justifyContent: 'center',
    alignItems: 'center',
    borderRadius: 6,
    borderColor: '#2e4453',
    backgroundColor: '#2e4453'
  },
  subscriptInactive: {
    color: '#7b9ab1',
    fontWeight: 'bold',
    fontSize: 13,
    alignSelf: 'flex-end',
    marginLeft: marginLeft,
    marginBottom: marginBottom
  },
  subscriptInactiveDark: {
    color: '#a7aaad' // $gray_20

  },
  subscriptActive: {
    color: 'white',
    fontWeight: 'bold',
    fontSize: 13,
    alignSelf: 'flex-end',
    marginLeft: marginLeft,
    marginBottom: marginBottom
  }
});
export function Button(props) {
  var children = props.children,
      onClick = props.onClick,
      disabled = props.disabled,
      hint = props.hint,
      _props$fixedRatio = props.fixedRatio,
      fixedRatio = _props$fixedRatio === void 0 ? true : _props$fixedRatio,
      getStylesFromColorScheme = props.getStylesFromColorScheme,
      ariaDisabled = props['aria-disabled'],
      ariaLabel = props['aria-label'],
      ariaPressed = props['aria-pressed'],
      subscript = props['data-subscript'];
  var isDisabled = ariaDisabled || disabled;

  var buttonViewStyle = _objectSpread({
    opacity: isDisabled ? 0.3 : 1
  }, fixedRatio && styles.fixedRatio, ariaPressed ? styles.buttonActive : styles.buttonInactive);

  var states = [];

  if (ariaPressed) {
    states.push('selected');
  }

  if (isDisabled) {
    states.push('disabled');
  }

  var subscriptInactive = getStylesFromColorScheme(styles.subscriptInactive, styles.subscriptInactiveDark);
  var newChildren = Children.map(children, function (child) {
    return child ? cloneElement(child, {
      colorScheme: props.preferredColorScheme,
      __unstableActive: ariaPressed
    }) : child;
  });
  return createElement(TouchableOpacity, {
    activeOpacity: 0.7,
    accessible: true,
    accessibilityLabel: ariaLabel,
    accessibilityStates: states,
    accessibilityRole: 'button',
    accessibilityHint: hint,
    onPress: onClick,
    style: styles.container,
    disabled: isDisabled
  }, createElement(View, {
    style: buttonViewStyle
  }, createElement(View, {
    style: {
      flexDirection: 'row'
    }
  }, newChildren, subscript && createElement(Text, {
    style: ariaPressed ? styles.subscriptActive : subscriptInactive
  }, subscript))));
}
export default withPreferredColorScheme(Button);
