// @wordpress/block-library/build-module/image/edit.native.js
import _extends from "@babel/runtime/helpers/esm/extends";
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _possibleConstructorReturn from "@babel/runtime/helpers/esm/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/esm/getPrototypeOf";
import _assertThisInitialized from "@babel/runtime/helpers/esm/assertThisInitialized";
import _inherits from "@babel/runtime/helpers/esm/inherits";
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";

var _sizeOptionLabels;

import { createElement } from "@wordpress/element";

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

/**
 * External dependencies
 */
import React from 'react';
import { View, ImageBackground, Text, TouchableWithoutFeedback, Dimensions } from 'react-native';
import { requestMediaImport, mediaUploadSync, requestImageFailedRetryDialog, requestImageUploadCancelDialog, requestImageFullscreenPreview, showMediaEditorButton } from 'react-native-gutenberg-bridge';
import { isEmpty, map, get } from 'lodash';
/**
 * WordPress dependencies
 */

import { Icon, PanelBody, PanelActions, SelectControl, TextControl, ToggleControl, ToolbarButton, ToolbarGroup } from '@wordpress/components';
import { BlockCaption, MediaPlaceholder, MediaUpload, MediaUploadProgress, MEDIA_TYPE_IMAGE, BlockControls, InspectorControls, BlockAlignmentToolbar, MediaEdit } from '@wordpress/block-editor';
import { __, sprintf } from '@wordpress/i18n';
import { isURL } from '@wordpress/url';
import { doAction, hasAction } from '@wordpress/hooks';
import { compose, withPreferredColorScheme } from '@wordpress/compose';
import { withSelect } from '@wordpress/data';
import { image as icon } from '@wordpress/icons';
/**
 * Internal dependencies
 */

import styles from './styles.scss';
import { editImageIcon } from './icon';
import SvgIconRetry from './icon-retry';
import SvgIconCustomize from './icon-customize';
import { getUpdatedLinkTargetSettings } from './utils';
import { LINK_DESTINATION_CUSTOM, LINK_DESTINATION_NONE } from './constants';
var ICON_TYPE = {
  PLACEHOLDER: 'placeholder',
  RETRY: 'retry',
  UPLOAD: 'upload'
};
var IMAGE_SIZE_THUMBNAIL = 'thumbnail';
var IMAGE_SIZE_MEDIUM = 'medium';
var IMAGE_SIZE_LARGE = 'large';
var IMAGE_SIZE_FULL_SIZE = 'full';
var DEFAULT_SIZE_SLUG = IMAGE_SIZE_LARGE;
var sizeOptionLabels = (_sizeOptionLabels = {}, _defineProperty(_sizeOptionLabels, IMAGE_SIZE_THUMBNAIL, __('Thumbnail')), _defineProperty(_sizeOptionLabels, IMAGE_SIZE_MEDIUM, __('Medium')), _defineProperty(_sizeOptionLabels, IMAGE_SIZE_LARGE, __('Large')), _defineProperty(_sizeOptionLabels, IMAGE_SIZE_FULL_SIZE, __('Full Size')), _sizeOptionLabels);
var sizeOptions = map(sizeOptionLabels, function (label, option) {
  return {
    value: option,
    label: label
  };
}); // Default Image ratio 4:3

var IMAGE_ASPECT_RATIO = 4 / 3;

var getUrlForSlug = function getUrlForSlug(image, _ref) {
  var sizeSlug = _ref.sizeSlug;
  return get(image, ['media_details', 'sizes', sizeSlug, 'source_url']);
};

export var ImageEdit =
/*#__PURE__*/
function (_React$Component) {
  _inherits(ImageEdit, _React$Component);

  function ImageEdit(props) {
    var _this;

    _classCallCheck(this, ImageEdit);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(ImageEdit).call(this, props));
    _this.state = {
      isCaptionSelected: false
    };
    _this.finishMediaUploadWithSuccess = _this.finishMediaUploadWithSuccess.bind(_assertThisInitialized(_this));
    _this.finishMediaUploadWithFailure = _this.finishMediaUploadWithFailure.bind(_assertThisInitialized(_this));
    _this.mediaUploadStateReset = _this.mediaUploadStateReset.bind(_assertThisInitialized(_this));
    _this.onSelectMediaUploadOption = _this.onSelectMediaUploadOption.bind(_assertThisInitialized(_this));
    _this.updateMediaProgress = _this.updateMediaProgress.bind(_assertThisInitialized(_this));
    _this.updateAlt = _this.updateAlt.bind(_assertThisInitialized(_this));
    _this.updateImageURL = _this.updateImageURL.bind(_assertThisInitialized(_this));
    _this.onSetLinkDestination = _this.onSetLinkDestination.bind(_assertThisInitialized(_this));
    _this.onSetNewTab = _this.onSetNewTab.bind(_assertThisInitialized(_this));
    _this.onSetSizeSlug = _this.onSetSizeSlug.bind(_assertThisInitialized(_this));
    _this.onImagePressed = _this.onImagePressed.bind(_assertThisInitialized(_this));
    _this.onClearSettings = _this.onClearSettings.bind(_assertThisInitialized(_this));
    _this.onFocusCaption = _this.onFocusCaption.bind(_assertThisInitialized(_this));
    _this.updateAlignment = _this.updateAlignment.bind(_assertThisInitialized(_this));
    return _this;
  }

  _createClass(ImageEdit, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      var _this$props = this.props,
          attributes = _this$props.attributes,
          setAttributes = _this$props.setAttributes; // This will warn when we have `id` defined, while `url` is undefined.
      // This may help track this issue: https://github.com/wordpress-mobile/WordPress-Android/issues/9768
      // where a cancelled image upload was resulting in a subsequent crash.

      if (attributes.id && !attributes.url) {
        // eslint-disable-next-line no-console
        console.warn('Attributes has id with no url.');
      } // Detect any pasted image and start an upload


      if (!attributes.id && attributes.url && attributes.url.indexOf('file:') === 0) {
        requestMediaImport(attributes.url, function (id, url) {
          if (url) {
            setAttributes({
              id: id,
              url: url
            });
          }
        });
      } // Make sure we mark any temporary images as failed if they failed while
      // the editor wasn't open


      if (attributes.id && attributes.url && !isURL(attributes.url)) {
        mediaUploadSync();
      }
    }
  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      // this action will only exist if the user pressed the trash button on the block holder
      if (hasAction('blocks.onRemoveBlockCheckUpload') && this.state.isUploadInProgress) {
        doAction('blocks.onRemoveBlockCheckUpload', this.props.attributes.id);
      }
    }
  }, {
    key: "componentDidUpdate",
    value: function componentDidUpdate(previousProps) {
      if (!previousProps.image && this.props.image) {
        var _this$props2 = this.props,
            image = _this$props2.image,
            attributes = _this$props2.attributes;
        var url = getUrlForSlug(image, attributes) || image.source_url;
        this.props.setAttributes({
          url: url
        });
      }
    }
  }, {
    key: "onImagePressed",
    value: function onImagePressed() {
      var _this$props3 = this.props,
          attributes = _this$props3.attributes,
          image = _this$props3.image;

      if (this.state.isUploadInProgress) {
        requestImageUploadCancelDialog(attributes.id);
      } else if (attributes.id && !isURL(attributes.url)) {
        requestImageFailedRetryDialog(attributes.id);
      } else if (!this.state.isCaptionSelected) {
        requestImageFullscreenPreview(attributes.url, image && image.source_url);
      }

      this.setState({
        isCaptionSelected: false
      });
    }
  }, {
    key: "updateMediaProgress",
    value: function updateMediaProgress(payload) {
      var setAttributes = this.props.setAttributes;

      if (payload.mediaUrl) {
        setAttributes({
          url: payload.mediaUrl
        });
      }

      if (!this.state.isUploadInProgress) {
        this.setState({
          isUploadInProgress: true
        });
      }
    }
  }, {
    key: "finishMediaUploadWithSuccess",
    value: function finishMediaUploadWithSuccess(payload) {
      var setAttributes = this.props.setAttributes;
      setAttributes({
        url: payload.mediaUrl,
        id: payload.mediaServerId
      });
      this.setState({
        isUploadInProgress: false
      });
    }
  }, {
    key: "finishMediaUploadWithFailure",
    value: function finishMediaUploadWithFailure(payload) {
      var setAttributes = this.props.setAttributes;
      setAttributes({
        id: payload.mediaId
      });
      this.setState({
        isUploadInProgress: false
      });
    }
  }, {
    key: "mediaUploadStateReset",
    value: function mediaUploadStateReset() {
      var setAttributes = this.props.setAttributes;
      setAttributes({
        id: null,
        url: null
      });
      this.setState({
        isUploadInProgress: false
      });
    }
  }, {
    key: "updateAlt",
    value: function updateAlt(newAlt) {
      this.props.setAttributes({
        alt: newAlt
      });
    }
  }, {
    key: "updateImageURL",
    value: function updateImageURL(url) {
      this.props.setAttributes({
        url: url,
        width: undefined,
        height: undefined
      });
    }
  }, {
    key: "updateAlignment",
    value: function updateAlignment(nextAlign) {
      this.props.setAttributes({
        align: nextAlign
      });
    }
  }, {
    key: "onSetLinkDestination",
    value: function onSetLinkDestination(href) {
      this.props.setAttributes({
        linkDestination: LINK_DESTINATION_CUSTOM,
        href: href
      });
    }
  }, {
    key: "onSetNewTab",
    value: function onSetNewTab(value) {
      var updatedLinkTarget = getUpdatedLinkTargetSettings(value, this.props.attributes);
      this.props.setAttributes(updatedLinkTarget);
    }
  }, {
    key: "onSetSizeSlug",
    value: function onSetSizeSlug(sizeSlug) {
      var image = this.props.image;
      var url = getUrlForSlug(image, {
        sizeSlug: sizeSlug
      });

      if (!url) {
        return null;
      }

      this.props.setAttributes({
        url: url,
        width: undefined,
        height: undefined,
        sizeSlug: sizeSlug
      });
    }
  }, {
    key: "onClearSettings",
    value: function onClearSettings() {
      this.props.setAttributes({
        alt: '',
        linkDestination: LINK_DESTINATION_NONE,
        href: undefined,
        linkTarget: undefined,
        sizeSlug: DEFAULT_SIZE_SLUG,
        rel: undefined
      });
    }
  }, {
    key: "onSelectMediaUploadOption",
    value: function onSelectMediaUploadOption(media) {
      var _this$props$attribute = this.props.attributes,
          id = _this$props$attribute.id,
          url = _this$props$attribute.url;
      var mediaAttributes = {
        id: media.id,
        url: media.url
      };
      var additionalAttributes; // Reset the dimension attributes if changing to a different image.

      if (!media.id || media.id !== id) {
        additionalAttributes = {
          width: undefined,
          height: undefined,
          sizeSlug: DEFAULT_SIZE_SLUG
        };
      } else {
        // Keep the same url when selecting the same file, so "Image Size" option is not changed.
        additionalAttributes = {
          url: url
        };
      }

      this.props.setAttributes(_objectSpread({}, mediaAttributes, {}, additionalAttributes));
    }
  }, {
    key: "onFocusCaption",
    value: function onFocusCaption() {
      if (this.props.onFocus) {
        this.props.onFocus();
      }

      if (!this.state.isCaptionSelected) {
        this.setState({
          isCaptionSelected: true
        });
      }
    }
  }, {
    key: "getIcon",
    value: function getIcon(iconType) {
      var iconStyle;

      switch (iconType) {
        case ICON_TYPE.RETRY:
          return createElement(Icon, _extends({
            icon: SvgIconRetry
          }, styles.iconRetry));

        case ICON_TYPE.PLACEHOLDER:
          iconStyle = this.props.getStylesFromColorScheme(styles.iconPlaceholder, styles.iconPlaceholderDark);
          break;

        case ICON_TYPE.UPLOAD:
          iconStyle = this.props.getStylesFromColorScheme(styles.iconUpload, styles.iconUploadDark);
          break;
      }

      return createElement(Icon, _extends({
        icon: icon
      }, iconStyle));
    }
  }, {
    key: "render",
    value: function render() {
      var _this2 = this;

      var _this$props4 = this.props,
          attributes = _this$props4.attributes,
          isSelected = _this$props4.isSelected,
          image = _this$props4.image;
      var align = attributes.align,
          url = attributes.url,
          height = attributes.height,
          width = attributes.width,
          alt = attributes.alt,
          href = attributes.href,
          id = attributes.id,
          linkTarget = attributes.linkTarget,
          sizeSlug = attributes.sizeSlug;
      var actions = [{
        label: __('Clear All Settings'),
        onPress: this.onClearSettings
      }];

      var getToolbarEditButton = function getToolbarEditButton(open) {
        return createElement(BlockControls, null, createElement(ToolbarGroup, null, createElement(ToolbarButton, {
          title: __('Edit image'),
          icon: editImageIcon,
          onClick: open
        })), createElement(BlockAlignmentToolbar, {
          value: align,
          onChange: _this2.updateAlignment,
          isCollapsed: false
        }));
      };

      var getInspectorControls = function getInspectorControls() {
        return createElement(InspectorControls, null, createElement(PanelBody, {
          title: __('Image settings')
        }, createElement(TextControl, {
          icon: 'admin-links',
          label: __('Link To'),
          value: href || '',
          valuePlaceholder: __('Add URL'),
          onChange: _this2.onSetLinkDestination,
          autoCapitalize: "none",
          autoCorrect: false,
          keyboardType: "url"
        }), createElement(ToggleControl, {
          icon: 'external',
          label: __('Open in new tab'),
          checked: linkTarget === '_blank',
          onChange: _this2.onSetNewTab
        }), // eslint-disable-next-line no-undef
        image && __DEV__ && createElement(SelectControl, {
          hideCancelButton: true,
          icon: 'editor-expand',
          label: __('Size'),
          value: sizeOptionLabels[sizeSlug || DEFAULT_SIZE_SLUG],
          onChangeValue: function onChangeValue(newValue) {
            return _this2.onSetSizeSlug(newValue);
          },
          options: sizeOptions
        }), createElement(TextControl, {
          icon: 'editor-textcolor',
          label: __('Alt Text'),
          value: alt || '',
          valuePlaceholder: __('None'),
          separatorType: 'none',
          onChangeValue: _this2.updateAlt
        })), createElement(PanelActions, {
          actions: actions
        }));
      };

      if (!url) {
        return createElement(View, {
          style: {
            flex: 1
          }
        }, createElement(MediaPlaceholder, {
          allowedTypes: [MEDIA_TYPE_IMAGE],
          onSelect: this.onSelectMediaUploadOption,
          icon: this.getIcon(ICON_TYPE.PLACEHOLDER),
          onFocus: this.props.onFocus
        }));
      }

      var alignToFlex = {
        left: 'flex-start',
        center: 'center',
        right: 'flex-end',
        full: 'center',
        wide: 'center'
      };
      var imageContainerHeight = Dimensions.get('window').width / IMAGE_ASPECT_RATIO;

      var editImageComponent = function editImageComponent(_ref2) {
        var open = _ref2.open,
            mediaOptions = _ref2.mediaOptions;
        return createElement(TouchableWithoutFeedback, {
          onPress: open
        }, createElement(View, {
          style: styles.editContainer
        }, createElement(View, {
          style: styles.edit
        }, mediaOptions(), createElement(Icon, _extends({
          size: 16,
          icon: SvgIconCustomize
        }, styles.iconCustomise)))));
      };

      var getImageComponent = function getImageComponent(openMediaOptions, getMediaOptions) {
        return createElement(TouchableWithoutFeedback, {
          accessible: !isSelected,
          onPress: _this2.onImagePressed,
          onLongPress: openMediaOptions,
          disabled: !isSelected
        }, createElement(View, {
          style: {
            flex: 1
          }
        }, getInspectorControls(), getMediaOptions(), !_this2.state.isCaptionSelected && getToolbarEditButton(openMediaOptions), createElement(MediaUploadProgress, {
          height: height,
          width: width,
          coverUrl: url,
          mediaId: id,
          onUpdateMediaProgress: _this2.updateMediaProgress,
          onFinishMediaUploadWithSuccess: _this2.finishMediaUploadWithSuccess,
          onFinishMediaUploadWithFailure: _this2.finishMediaUploadWithFailure,
          onMediaUploadStateReset: _this2.mediaUploadStateReset,
          renderContent: function renderContent(_ref3) {
            var isUploadInProgress = _ref3.isUploadInProgress,
                isUploadFailed = _ref3.isUploadFailed,
                finalWidth = _ref3.finalWidth,
                finalHeight = _ref3.finalHeight,
                imageWidthWithinContainer = _ref3.imageWidthWithinContainer,
                retryMessage = _ref3.retryMessage;
            var opacity = isUploadInProgress ? 0.3 : 1;
            var imageBorderOnSelectedStyle = isSelected && !(isUploadInProgress || isUploadFailed || _this2.state.isCaptionSelected) ? styles.imageBorder : '';
            var iconRetryContainer = createElement(View, {
              style: styles.modalIcon
            }, _this2.getIcon(ICON_TYPE.RETRY));
            return createElement(View, {
              style: {
                flex: 1,
                // only set alignSelf if an image exists because alignSelf causes the placeholder
                // to disappear when an aligned image can't be downloaded
                // https://github.com/wordpress-mobile/gutenberg-mobile/issues/1592
                alignSelf: imageWidthWithinContainer && alignToFlex[align]
              }
            }, !imageWidthWithinContainer && createElement(View, {
              style: [_this2.props.getStylesFromColorScheme(styles.imageContainerUpload, styles.imageContainerUploadDark), {
                height: imageContainerHeight
              }]
            }, createElement(View, {
              style: styles.imageUploadingIconContainer
            }, _this2.getIcon(ICON_TYPE.UPLOAD))), createElement(ImageBackground, {
              accessible: true,
              disabled: !isSelected,
              accessibilityLabel: alt,
              accessibilityHint: __('Double tap and hold to edit'),
              accessibilityRole: 'imagebutton',
              style: [imageBorderOnSelectedStyle, {
                width: finalWidth,
                height: finalHeight,
                opacity: opacity
              }],
              resizeMethod: "scale",
              source: {
                uri: url
              },
              key: url
            }, isUploadFailed && createElement(View, {
              style: [styles.imageContainer, {
                flex: 1,
                backgroundColor: 'rgba(0, 0, 0, 0.5)'
              }]
            }, iconRetryContainer, createElement(Text, {
              style: styles.uploadFailedText
            }, retryMessage)), isSelected && !isUploadInProgress && !isUploadFailed && finalWidth && finalHeight && showMediaEditorButton && createElement(MediaEdit, {
              allowedTypes: [MEDIA_TYPE_IMAGE],
              onSelect: _this2.onSelectMediaUploadOption,
              source: {
                uri: url
              },
              openReplaceMediaOptions: openMediaOptions,
              render: editImageComponent
            })));
          }
        }), createElement(BlockCaption, {
          clientId: _this2.props.clientId,
          isSelected: _this2.state.isCaptionSelected,
          accessible: true,
          accessibilityLabelCreator: function accessibilityLabelCreator(caption) {
            return isEmpty(caption) ?
            /* translators: accessibility text. Empty image caption. */
            'Image caption. Empty' : sprintf(
            /* translators: accessibility text. %s: image caption. */
            __('Image caption. %s'), caption);
          },
          onFocus: _this2.onFocusCaption,
          onBlur: _this2.props.onBlur // always assign onBlur as props

        })));
      };

      return createElement(MediaUpload, {
        allowedTypes: [MEDIA_TYPE_IMAGE],
        onSelect: this.onSelectMediaUploadOption,
        render: function render(_ref4) {
          var open = _ref4.open,
              getMediaOptions = _ref4.getMediaOptions;
          return getImageComponent(open, getMediaOptions);
        }
      });
    }
  }], [{
    key: "getDerivedStateFromProps",
    value: function getDerivedStateFromProps(props, state) {
      // Avoid a UI flicker in the toolbar by insuring that isCaptionSelected
      // is updated immediately any time the isSelected prop becomes false
      return {
        isCaptionSelected: props.isSelected && state.isCaptionSelected
      };
    }
  }]);

  return ImageEdit;
}(React.Component);
export default compose([withSelect(function (select, props) {
  var _select = select('core'),
      getMedia = _select.getMedia;

  var id = props.attributes.id,
      isSelected = props.isSelected;
  return {
    image: id && isSelected ? getMedia(id) : null
  };
}), withPreferredColorScheme])(ImageEdit);
