// @wordpress/components/build/button/index.native.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Button = Button;
exports.default = void 0;

var _element = require("@wordpress/element");

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread"));

var _reactNative = require("react-native");

var _compose = require("@wordpress/compose");

/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */
var isAndroid = _reactNative.Platform.OS === 'android';
var marginBottom = isAndroid ? -0.5 : 0;
var marginLeft = -3;

var styles = _reactNative.StyleSheet.create({
  container: {
    flex: 1,
    padding: 3,
    justifyContent: 'center',
    alignItems: 'center'
  },
  buttonInactive: {
    flex: 1,
    flexDirection: 'row',
    justifyContent: 'center',
    alignItems: 'center'
  },
  fixedRatio: {
    aspectRatio: 1
  },
  buttonActive: {
    flex: 1,
    flexDirection: 'row',
    justifyContent: 'center',
    alignItems: 'center',
    borderRadius: 6,
    borderColor: '#2e4453',
    backgroundColor: '#2e4453'
  },
  subscriptInactive: {
    color: '#7b9ab1',
    fontWeight: 'bold',
    fontSize: 13,
    alignSelf: 'flex-end',
    marginLeft: marginLeft,
    marginBottom: marginBottom
  },
  subscriptInactiveDark: {
    color: '#a7aaad' // $gray_20

  },
  subscriptActive: {
    color: 'white',
    fontWeight: 'bold',
    fontSize: 13,
    alignSelf: 'flex-end',
    marginLeft: marginLeft,
    marginBottom: marginBottom
  }
});

function Button(props) {
  var children = props.children,
      onClick = props.onClick,
      disabled = props.disabled,
      hint = props.hint,
      _props$fixedRatio = props.fixedRatio,
      fixedRatio = _props$fixedRatio === void 0 ? true : _props$fixedRatio,
      getStylesFromColorScheme = props.getStylesFromColorScheme,
      ariaDisabled = props['aria-disabled'],
      ariaLabel = props['aria-label'],
      ariaPressed = props['aria-pressed'],
      subscript = props['data-subscript'];
  var isDisabled = ariaDisabled || disabled;
  var buttonViewStyle = (0, _objectSpread2.default)({
    opacity: isDisabled ? 0.3 : 1
  }, fixedRatio && styles.fixedRatio, ariaPressed ? styles.buttonActive : styles.buttonInactive);
  var states = [];

  if (ariaPressed) {
    states.push('selected');
  }

  if (isDisabled) {
    states.push('disabled');
  }

  var subscriptInactive = getStylesFromColorScheme(styles.subscriptInactive, styles.subscriptInactiveDark);

  var newChildren = _element.Children.map(children, function (child) {
    return child ? (0, _element.cloneElement)(child, {
      colorScheme: props.preferredColorScheme,
      __unstableActive: ariaPressed
    }) : child;
  });

  return (0, _element.createElement)(_reactNative.TouchableOpacity, {
    activeOpacity: 0.7,
    accessible: true,
    accessibilityLabel: ariaLabel,
    accessibilityStates: states,
    accessibilityRole: 'button',
    accessibilityHint: hint,
    onPress: onClick,
    style: styles.container,
    disabled: isDisabled
  }, (0, _element.createElement)(_reactNative.View, {
    style: buttonViewStyle
  }, (0, _element.createElement)(_reactNative.View, {
    style: {
      flexDirection: 'row'
    }
  }, newChildren, subscript && (0, _element.createElement)(_reactNative.Text, {
    style: ariaPressed ? styles.subscriptActive : subscriptInactive
  }, subscript))));
}

var _default = (0, _compose.withPreferredColorScheme)(Button);

exports.default = _default;
