// @wordpress/block-library/build/embed/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.others = exports.common = exports.settings = exports.name = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _coreEmbeds = require("./core-embeds");

var _icons = require("./icons");

var _settings = require("./settings");

var _i18n = require("@wordpress/i18n");

var _blocks = require("@wordpress/blocks");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

var name = 'core/embed';
exports.name = name;
var settings = (0, _settings.getEmbedBlockSettings)({
  title: (0, _i18n._x)('Embed', 'block title'),
  description: (0, _i18n.__)('Embed videos, images, tweets, audio, and other content from external sources.'),
  icon: _icons.embedContentIcon,
  // Unknown embeds should not be responsive by default.
  responsive: false,
  transforms: {
    from: [{
      type: 'raw',
      isMatch: function isMatch(node) {
        return node.nodeName === 'P' && /^\s*(https?:\/\/\S+)\s*$/i.test(node.textContent);
      },
      transform: function transform(node) {
        return (0, _blocks.createBlock)('core/embed', {
          url: node.textContent.trim()
        });
      }
    }]
  }
});
exports.settings = settings;

var common = _coreEmbeds.common.map(function (embedDefinition) {
  return _objectSpread({}, embedDefinition, {
    settings: (0, _settings.getEmbedBlockSettings)(embedDefinition.settings)
  });
});

exports.common = common;

var others = _coreEmbeds.others.map(function (embedDefinition) {
  return _objectSpread({}, embedDefinition, {
    settings: (0, _settings.getEmbedBlockSettings)(embedDefinition.settings)
  });
});

exports.others = others;
